llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -8701,95 +8734,341 @@ SDValue SystemZTargetLowering::combineSETCC(
+      return {};
+
+    int CCValidVal = CCValid->getZExtValue();
+    int CCMaskVal = CCMask->getZExtValue();
+    const auto &&TrueSDVals = simplifyAssumingCCVal(TrueVal, CC, DCI);
+    const auto &&FalseSDVals = simplifyAssumingCCVal(FalseVal, CC, DCI);
+    if (TrueSDVals.empty() || FalseSDVals.empty())
+      return {};
+    SDValue Op4CCReg = N->getOperand(4);
+    auto *Op4CCNode = Op4CCReg.getNode();
+    if (Op4CCNode && Op4CCNode != CCNode)
+      combineCCMask(Op4CCReg, CCValidVal, CCMaskVal, DCI);
+    Op4CCNode = Op4CCReg.getNode();
+    if (!Op4CCNode || Op4CCNode != CCNode)
+      return {};
+    SmallVector<SDValue, 4> MergedSDVals;
+    for (auto &CCVal : {0, 1, 2, 3})
+      MergedSDVals.emplace_back((((CCMaskVal & (1 << (3 - CCVal))) != 0) &&
+                                 ((CCValidVal & (1 << (3 - CCVal))) != 0))
+                                    ? TrueSDVals[CCVal]
+                                    : FalseSDVals[CCVal]);
+    return MergedSDVals;
+  }
+  SDValue Op0, Op1;
+  unsigned Opcode;
+  if (isValidBinaryOperation(Val, Op0, Op1, Opcode)) {
+    const auto &&Op0SDVals = simplifyAssumingCCVal(Op0, CC, DCI);
+    const auto &&Op1SDVals = simplifyAssumingCCVal(Op1, CC, DCI);
+    if (Op0SDVals.empty() || Op1SDVals.empty())
+      return {};
+    SmallVector<SDValue, 4> BinaryOpSDVals;
+    for (auto CCVal : {0, 1, 2, 3})
+      BinaryOpSDVals.emplace_back(DAG.getNode(
+          Opcode, DL, Val.getValueType(), Op0SDVals[CCVal], Op1SDVals[CCVal]));
+    return BinaryOpSDVals;
+  }
+  return {};
+}
+
+bool SystemZTargetLowering::combineCCMask(SDValue &CCReg, int &CCValid,
+                                          int &CCMask,
+                                          DAGCombinerInfo &DCI) const {
   // We have a SELECT_CCMASK or BR_CCMASK comparing the condition code
   // set by the CCReg instruction using the CCValid / CCMask masks,
-  // If the CCReg instruction is itself a ICMP testing the condition
+  // If the CCReg instruction is itself a ICMP / TM  testing the condition
   // code set by some other instruction, see whether we can directly
   // use that condition code.
-
-  // Verify that we have an ICMP against some constant.
-  if (CCValid != SystemZ::CCMASK_ICMP)
-    return false;
-  auto *ICmp = CCReg.getNode();
-  if (ICmp->getOpcode() != SystemZISD::ICMP)
-    return false;
-  auto *CompareLHS = ICmp->getOperand(0).getNode();
-  auto *CompareRHS = dyn_cast<ConstantSDNode>(ICmp->getOperand(1));
-  if (!CompareRHS)
+  auto *CCNode = CCReg.getNode();
+  if (!CCNode)
     return false;
+  const auto getConstFromConstSDVals = [](const SmallVector<SDValue, 4> &Vals) {
+    SmallVector<int, 4> CCVals;
+    for (const auto &Val : Vals)
+      if (auto *ConstNode = dyn_cast<ConstantSDNode>(Val.getNode()))
+        CCVals.emplace_back(ConstNode->getZExtValue());
+      else
+        return SmallVector<int, 4>();
+    return CCVals;
+  };
+  const auto getMSBPosSet = [](unsigned int Mask) {
```

Comment:
I don't think you need to implement this by hand, there's common LLVM helpers like `countl_zero` to do this.  (These helpers also work directly on `APInt`).